### DISCUSSION TITLE
[Implemented] Multi-pannel plots

### DISCUSSION THREAD

**Discussion by PavelBal** [Open-Systems-Pharmacology/OSPSuite-R#856] [Ideas]
It should be possible to combine multiple plots in one multi-pannel figure:

![image](https://user-images.githubusercontent.com/25903040/160800383-4a157b3f-1af0-4fbd-9f7e-1ace233f09dc.png)

I was trying to figure out how the API should look like using the currently proposed functional approach (`plotIndividualTimeProfile(dataCombined, plotConfiguration`) and I cannot come up with any idea.

Here how it is implemented in `esqlabsR`:

Simulated and observed data are grouped in an object of the class `DataMapping`. As opposed to `DataCombined`, `DataMapping` does not only combines data and defines groupings, but also implements a plotting method and defines some plot properties:

- x-axis limits (can be automatically calculated)
- y-axis limits (can be automatically calculated)
- x-axis lable (default to `Dimension [unit]`)
- y-axis lable (default to `Dimension [unit]`)
- x/y log
- Title
- Show legend TRUE/FALSE
- legend position
- x units
- y units
- `plotType` - individual profile, population profile, simVsObs
- populationQuantiles for population profile
- For the data sets:
  - Line types (can be automatically calculated)
  - symbol types (can be automatically calculated)
  - colors (can be automatically calculated)

So basically `DataMapping` is a combination of `DataCombined` with a plot configuration. Calling `DataMapping$plot()` creates a figure as defined by the `plotType`.

To create a multipanel figure, multiple `DataMapping` objects are passed to the function `plotMultiPanel`

`  plotMultiPanel(c(dataMapping_timeProfile, dataMapping_simVsObs), plotConfiguration)`

In this example, `dataMapping_timeProfile` has the `plotType` IndividualProfile, and `dataMapping_simVsObs` has the `plotType` simulated vs observed. `plotConfiguration` is an object that defines the output properties:

- output device (default device, png, or pdf)
- output file name
- output folder
- widht/height
- resolution
-  point size
- nr of colums in a multi-pannel (calculated automatically if not specified)

Will be happy to hear your thoughts on how to implement it in ospsuite-r. @IndrajeetPatil @Yuri05 @msevestre

**Comment by Yuri05** [comment]
I don't understand this spec at all. Why don't we just say:
* we create N plots, each of them with its own configuration
* and then just combine them into a plot grid in a "standard" way, e.g. https://cran.r-project.org/web/packages/egg/vignettes/Ecosystem.html 
?

So we could combine any types of plots (Individual time profiles, populatio time profiles, sensitivity plots, Gof plots, ddi ratio plots, custom plots, ...) in any order / layout.

![grafik](https://user-images.githubusercontent.com/25061876/161216526-abf9781f-2b97-4aad-a563-beb96f9131a1.png)

**Comment by PavelBal** [comment]
I was not aware of the `grid.arrange`, so I think we are good. Might want to think if we want an additional wrapper around it, but it's low prio now.

More and more I get a feeling I was re-implementing ggplot xD

**Comment by PavelBal** [comment]
OK, how do I create the following plot?

![image](https://user-images.githubusercontent.com/25903040/168045085-81e17645-a436-4936-9338-49865f9764e5.png)


I have three sub-plots, and want to arrange them. I also need to add title (SteadyState), and panel numbering (a, b, c)

**Comment by PavelBal** [comment]
@IndrajeetPatil

**Comment by IndrajeetPatil** [comment]
``` r
library(tlf)

# create a list of plots
ls_plots <- list(
  # first plot
  plotBoxWhisker(mtcars,
                 dataMapping = BoxWhiskerDataMapping$new(x = "am", y = "wt"), outliers = FALSE
  ),
  # second plot
  plotBoxWhisker(ToothGrowth,
                 dataMapping = BoxWhiskerDataMapping$new(x = "supp", y = "len")
  ),
  # third plot
  plotHistogram(x = mtcars$wt)
)

# create a new instance of plot config class
plotGridObj <- PlotGridConfiguration$new(ls_plots)

plotGridObj$title <- "SteadyState"
plotGridObj$nRows <- 2L
plotGridObj$tagLevels <- "a"

plotGrid(plotGridObj)
```

![](https://i.imgur.com/pwbHhop.png)

<sup>Created on 2022-05-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by PavelBal** [comment]
This does not work with plots created with `plotIndividualTimeProfile`, so it has to be implemented.

**Comment by IndrajeetPatil** [comment]
Works just fine for me. 

Are you sure you are using the development version of _tlf_ with the latest commit? 

```r
library(ospsuite)
library(ospsuite.utils)
library(tlf)

# import observed data (will return a list of DataSet objects)
dataSet <- loadDataSetsFromExcel(
  xlsFilePath = "CompiledDataSet.xlsx",
  importerConfiguration = loadDataImporterConfiguration("ImporterConfiguration.xml")
)

# simulation results
sim <- importResultsFromCSV(
  simulation = loadSimulation("Stevens_2012_placebo_indiv_results.pkml"),
  filePaths = "Stevens_2012_placebo_indiv_results.csv"
)

# create a `DataCombined` object and add observed and simulated data
myDataComb <- DataCombined$new()
myDataComb$addDataSets(dataSet)
myDataComb$addSimulationResults(sim,
  quantitiesOrPaths = list(
    "Organism|Lumen|Stomach|Metformin|Gastric retention",
    "Organism|Lumen|Stomach|Metformin|Gastric retention distal",
    "Organism|Lumen|Stomach|Metformin|Gastric retention proximal"
  )
)

# set groupings
myDataComb$setGroups(
  list(
    "Stevens_2012_placebo.Placebo_distal",
    "Organism|Lumen|Stomach|Metformin|Gastric retention distal",
    "Stevens_2012_placebo.Placebo_proximal",
    "Organism|Lumen|Stomach|Metformin|Gastric retention proximal",
    "Stevens_2012_placebo.Placebo_total",
    "Organism|Lumen|Stomach|Metformin|Gastric retention"
  ),
  list(
    # group 1
    "Stevens 2012 solid distal",
    "Stevens 2012 solid distal",
    # group 2
    "Stevens 2012 solid proximal",
    "Stevens 2012 solid proximal",
    # group 3
    "Stevens 2012 solid total",
    "Stevens 2012 solid total"
  )
)

myPC <- DefaultPlotConfiguration$new()
myPC$yUnit <- "%"

p1 <- plotIndividualTimeProfile(myDataComb, myPC)

myPC$linesLinetype <- names(tlf::Linetypes)
p2 <- plotIndividualTimeProfile(myDataComb, myPC)

myPC$linesLinetype <- tlf::Linetypes$dashed
myPC$pointsShape <- "circle"
p3 <- plotIndividualTimeProfile(myDataComb, myPC)

ls_plots <- list(p1, p2, p3)

# create a new instance of plot config class
plotGridObj <- PlotGridConfiguration$new(ls_plots)

plotGridObj$title <- "SteadyState"
plotGridObj$nRows <- 2L
plotGridObj$tagLevels <- "a"

plotGrid(plotGridObj)
```


![pGrid](https://user-images.githubusercontent.com/11330453/168060506-1c743926-d799-4f7b-a9c8-f42448282820.png)

**Comment by PavelBal** [comment]
Ok I hate R's lists.

How can I dynamically add plot objects to a list?


```
plots <- list()
plot <- plotIndividualTimeProfile(dataCombined = currentDataCombined, defaultPlotConfiguration = plotConfigurationScenario)

plots <- c(plots, plot)
```

And it does not work any more.

**Reply by IndrajeetPatil** [reply]
You can use `append()`:

``` r
plots <- list()
plot <- "x"
plots <- append(plots, plot)
plots
#> [[1]]
#> [1] "x"
```

<sup>Created on 2022-05-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by PavelBal** [comment]
Nope,

```
>   plots <- list()
> plots <- append(plots, plot)
> length(plots)
[1] 11
```

Actually, `append` uses `c`, so no difference to expect.

**Reply by IndrajeetPatil** [reply]
Is dynamically updating the list not what you wanted? `append()` seems to do the trick:

``` r
plots <- list()
x <- letters[1:3]

for (elem in x) {
  plots <- append(plots, elem)
}

plots
#> [[1]]
#> [1] "a"
#> 
#> [[2]]
#> [1] "b"
#> 
#> [[3]]
#> [1] "c"
```

<sup>Created on 2022-05-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by PavelBal** [comment]
Just try it with your own example with the plots:
```

dataSet <- loadDataSetsFromExcel(
  xlsFilePath = "CompiledDataSet.xlsx",
  importerConfiguration = loadDataImporterConfiguration("ImporterConfiguration.xml")
)
# create a `DataCombined` object and add observed and simulated data
myDataComb <- DataCombined$new()
myDataComb$addDataSets(dataSet)
myPC <- DefaultPlotConfiguration$new()
myPC$yUnit <- "%"

p1 <- plotIndividualTimeProfile(myDataComb, myPC)

ls_plots <- list()
ls_plots <- append(ls_plots, p1)

plotGridObj <- PlotGridConfiguration$new(ls_plots)
plotGrid(plotGridObj)

> plotGrid(plotGridObj)
 Error: Only know how to add ggplots and/or grobs
3.
stop("Only know how to add ggplots and/or grobs", call. = FALSE)
2.
patchwork::wrap_plots(plotGridConfiguration$plotList, ncol = plotGridConfiguration$nColumns, 
nrow = plotGridConfiguration$nRows, byrow = plotGridConfiguration$byRow, 
widths = plotGridConfiguration$widths, heights = plotGridConfiguration$heights, 
guides = plotGridConfiguration$guides, design = plotGridConfiguration$design) at plot-grid.R#72
1.
plotGrid(plotGridObj)

```

so, no, it does not work.

**Reply by IndrajeetPatil** [reply]
That should fix it.

```r
ls_plots <- list()
ls_plots <- append(ls_plots, list(p1))
```

**Comment by PavelBal** [comment]
This is so ugly. Can we implement a method `addPlot` for `PlotGridConfiguration` class?

**Reply by msevestre** [reply]
yes!!! 
or not use list :)
